DB設計

miracletea opened the pull request:
# What
DB設計をおこないました。確認をお願いします。

# Why
フリマアプリ機能実装のため。

gyazoURL(ER図)
https://gyazo.com/0de10fd350a114e889c138fa65a27526

miracletea committed:
DB設計

div-reviewer reviewed (commented):
miracletea様、学習お疲れさまです。
DB設計の初回プルリクエスト提出ありがとうございました。

プルリクエストのWhatWhyの箇所とコメント欄を確認したのですが、`ER図のgyazoURL`が確認できませんでした。
要件表にございます通り、`「ER図」`のご提出をお願いできればと存じます。

追加ができましたら、googleフォームにて再度プルリクエストの御提出をお願いいたします。

**操作などについてご質問、ご不明点ございましたら、コネクトかチャットメンターにお問い合わせ下さいませ。**

div-reviewer reviewed (commented):
miracletea様、学習お疲れ様です。
再度DB設計のプルリクエストのご提出ありがとうございました。

内容を拝見しましたが、「https://github.com/miracletea/furima-39667/pull/12」のコメント欄に`ER図を撮影したGyazoのURL`が見当たりませんでした。
（GitHubの通信の可能性も考え、何度かリロードも試しましたが閲覧ができませんでした）

コメント欄にGyazoのURLがはれましたら、`再度googleフォームのご提出`をお願いいたします。

操作が難しい場合は、コネクト・チャットメンターに相談しながら行いましょう。
お手数ですがどうぞよろしくお願いいたします。

div-reviewer reviewed (commented):
gyazoURL(ER図)
https://gyazo.com/0de10fd350a114e889c138fa65a27526

上記の内容が確認取れたので、今からレビューの確認に入りたいと思います。
ご提出ありがとうございました。もう少々お待ちくださいませ。

div-reviewer reviewed (commented):
miracletea様、学習お疲れさまです。
DB設計の再レビュー依頼頂きありがとうございました。プルリクエスト確認させていただきました。

テーブル間のアソシエーションまでよく考えられて作成されていますね！素晴らしいです。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いているER図も修正範囲だった場合、次回プルリクエスト提出時には最新の画像の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。**

miracletea committed:
emailにオプションユニークを追加

| email               | string | null: false |
ユーザー管理機能の実装条件であるメールアドレスが一意性であること（unique: true）になるように、Optionsに記述を追記しましょう。
理由
・ユーザー管理機能で使用していただきますdeviseの設定で、emailカラムには一意性制約がかかっている為
・実装条件に「メールアドレスは一意性であること」とあること

miracletea committed:
text型からstring型に変更

| last_name          | text     | null: false |
| first_name         | text     | null: false |
| last_name_kana     | text     | null: false |
| first_name_kana    | text     | null: false |
Typeをstring型に変更しましょう。

名前は255文字で足りるかと思うためです。

・string型：文字は255字まで
・text型：文字を255文字より多く使用したい場合

miracletea committed:
誕生日のタイプを「date」型に変更

誕生日のタイプを「date」型にし、1つのカラムにまとめましょう。

理由

・'20190229"といった無効な日付データの保存を防ぐことが出来ます
・年、月、日を1つでまとめて保存することが出来ます
例

| birthday               | date | null: false

miracletea committed:
アソシエーションの削除

- has_one :shipping_address
上記は削除しましょう。
購入記録の中に、購入者・購入品が入っており、こちらと住所が繋がっていれば問題ないためです。

miracletea committed:
activehashを用いたカラム、型に修正

商品の状態, 配送料の負担, 発送元の地域,発送までの日数,カテゴリーのカラム

上記カラムはactivehashを用いた実装になるため、activehashのルールに沿ったカラム名と型にしましょう。
activehashにつきましては以下カリキュラムをご参考にいただければと思います。

Lesson1 最終課題補足カリキュラム
カテゴリーの選択を工夫しよう

miracletea committed:
住所テーブルの参照テーブルをuserからrecordに変更

住所テーブルにはユーザーではなく購入管理テーブルの外部キーを設定しましょう！

理由
購入管理テーブルの外部キーを設けることで「この商品は誰が購入したものなのか」を紐づけることができる為です。

miracletea committed:
state_province_idの追加

項目が一つ抜けているため、確認しましょう。

・郵便番号
・都道府県
・市区町村
・番地
・建物名
・電話番号
・購入管理テーブルの外部キー

miracletea committed:
電話番号をstring型に変更

電話番号はstring型で実装しましょう。

理由
integer（整数）型の場合、仮に「01234」を保存しようとすると先頭の0が読み込まれず「1234」に変換されてしまいます。電話番号は0からスタートするためstring（文字）型にしてあげましょう。

例：入力された番号　09012345678の場合
integer型の場合→9012345678
string型の場合→09012345678

miracletea committed:
住所テーブルと購入管理テーブルのアソシエーションの変更

今回の設計では1件の配送先住所に1件の購入管理が紐づく（1対1）実装になるかと思われます。
アソシエーションもそのように組んでみましょう。

■住所テーブルと購入管理テーブルの関係
・購入管理テーブル
　has_one :住所テーブル

・住所テーブル
　belongs_to :購入管理テーブル
``

miracletea committed:
単数形のアソシエーションのモデルを単数形に変更

- belongs_to :users
- belongs_to :items
@div-erika div-erika 1 hour ago
- belongs_to :users
- belongs_to :items
belongs_toは単数系のアソシエーションになります。
右側のモデル名も単数系にしましょう。

miracletea committed:
ER図の修正

miracletea commented:
修正しました。
ご確認、宜しくお願い致します。

gyazoURL(ER図)
https://gyazo.com/6fe9b6f05acfbed3a9fa5767a2d71b1f

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,70 @@
 
-* Configuration
+- has_many :items
+- has_one :record
```
```rb
- has_one :record
```
1人のユーザーが何度でも購入できるように`has_many :複数形`の形に変更しましょう。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| ship_form_id          | integer     | null: false                     |
+| days_to_ship_date_id  | integer     | null: false                     |
+| price                 | integer     | null: false                     |
+| user                  | reference   | null: false, foreign_key: true  |
```
```rb
| user | reference | null: false, foreign_key: true |
```
`reference`ではなく末尾にsをつけて`references`と記述するのが正しい形なので修正しましょう。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| Column             | Type        | Options                         |
+| ------------------ | ----------- | ------------------------------- |
+| postal_code        | string      | null: false                     |
+| state_province_id  | integer     | null: false                     |
```
```Rb
| state_province_id | integer | null: false |
```
都道府県のカラム名は商品の都道府県と同じにしましょう。
都道府県のファイルを実装する際に作成するファイルが１つで済むためです。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+
+### Association
+
```
外部キーがあるので、ユーザーと商品とのアソシエーションも記述しましょう。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
コメントを追記致しましたので、ご確認お願いします。
また、次回レビュー依頼時に最新状態のER図の添付もお願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

miracletea committed:
has_oneからhas_manyに修正

* Configuration
- has_many :items
- has_one :record
@TakashiKozuma TakashiKozuma 17 minutes ago
- has_one :record
1人のユーザーが何度でも購入できるようにhas_many :複数形の形に変更しましょう。

miracletea committed:
referenceの誤記を修正

| ship_form_id          | integer     | null: false                     |
| days_to_ship_date_id  | integer     | null: false                     |
| price                 | integer     | null: false                     |
| user                  | reference   | null: false, foreign_key: true  |
@TakashiKozuma TakashiKozuma 17 minutes ago
| user | reference | null: false, foreign_key: true |
referenceではなく末尾にsをつけてreferencesと記述するのが正しい形なので修正しましょう。

miracletea committed:
送信元、送信先のカラムを一つに統一

都道府県のカラム名は商品の都道府県と同じにしましょう。
都道府県のファイルを実装する際に作成するファイルが１つで済むためです。

miracletea committed:
ユーザーと商品のアソシエーション追加

外部キーがあるので、ユーザーと商品とのアソシエーションも記述しましょう。

miracletea commented:
ご確認ありがとうございます。
コメント内容を反映しました、ご確認よろしくお願い致します。

gyazoURL(ER図)
https://gyazo.com/0768e123bd3b09f6ebb50e4c06b55c0f

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
